Change Config.java to Enable SlimeFun-side Optimization

creator3 opened the pull request:
SlimeFun  does this a few times in `BlockStorage`
```java
new Config("data-storage/Slimefun/temp.yml");
```
Unfortunately, this still incurs the cost of opening a file. On my server, I'm running a version of SlimeFun that has a special subclass of Config that isn't backed by a `FileConfiguration`. Adding this constructor allows me to call
```java
super( (File) null, (FileConfiguration) null);
```
and avoid loading a .yml. I've overridden any methods that could cause a NullPointerException. This seems to have helped performance (like CargoNets), so I'm trying to PR it, and this is the first step. I accidentally ran an autoformatter, but I think I've gotten the diff back down to something reasonable. There is one spot I had to change "§" to "\u00a7" because I didn't want to deal with figuring out the correct encoding.






On the SlimeFun side, I have some questions that I'll just ask here.
First, is there a reason the tempfile is saved [on this line](https://github.com/TheBusyBiscuit/Slimefun4/blob/1d0660db535cbefc87ec1d5dd5863c450401ed3b/src/me/mrCookieSlime/Slimefun/api/BlockStorage.java#L229)? Second, is [this call](https://github.com/TheBusyBiscuit/Slimefun4/blob/1d0660db535cbefc87ec1d5dd5863c450401ed3b/src/me/mrCookieSlime/Slimefun/api/BlockStorage.java#L370) to `_integrated_removeBlockInfo` necessary? The reason I ask is because it isn't called in
```java
void setBlockInfo(Block b, String json, boolean updateTicker) 
```

Third, is it okay to cache the parsed string data?
```java
Map<Location, String> storage
```
would become
```java
Map<Location, Map_Str_Str_Config> storage
```
Or it could be cached but done lazily if you think doing all the parsing at startup would be too much.

Finally, can my memory-only Config be backed by a Map<String, String> and not support any of the more complex types supported by the regular Config? The current [parseJSON](https://github.com/TheBusyBiscuit/Slimefun4/blob/1d0660db535cbefc87ec1d5dd5863c450401ed3b/src/me/mrCookieSlime/Slimefun/api/BlockStorage.java#L299) is like this, and it seems best to keep similar behavior.

creator3 committed:
Add Config(File file, FileConfiguration config)

Reduce direct usage of this.config

TheBusyBiscuit commented:
Do not update the version in the pom.xml.
This is done automatically when a version is released.

TheBusyBiscuit commented on src/me/mrCookieSlime/CSCoreLibPlugin/Configuration/Config.java:
```diff
@@ -202,16 +217,16 @@ public ItemStack getItem(String path) {
+				if (this.contains(path + "_extra.custom-skullOwner") && !((ItemStack) item).getItemMeta().hasDisplayName()) {
 					ItemMeta im = ((ItemStack) item).getItemMeta();
-					im.setDisplayName("�r" + config.getString(path + "_extra.custom-skullOwner") + "'s Head");
+					im.setDisplayName("\u00a7r" + this.getString(path + "_extra.custom-skullOwner") + "'s Head");
```
Use ChatColor.translateAlternateColorCodes('&', ...) instead.

TheBusyBiscuit commented on pom.xml:
```diff
@@ -6,7 +6,7 @@
     <groupId>me.mrCookieSlime</groupId>
     <artifactId>CS-CoreLib</artifactId>
-    <version>1.5.16</version>
+    <version>1.5.17</version>
```
Do not change this manually.

TheBusyBiscuit commented on src/me/mrCookieSlime/CSCoreLibPlugin/Configuration/Config.java:
```diff
@@ -79,6 +90,10 @@ public FileConfiguration getConfiguration() {
 		return this.config;
 	}
 	
+	protected void setSimpleValue(String path, Object value) {
```
Just name it "store" ? Sounds better in my opinion.

TheBusyBiscuit commented:
@creator3 
"First, is there a reason the tempfile is saved on this line?"
Yes, because it isn't a temporary file, it's the chunks.sfc file.

"Second, is this call to _integrated_removeBlockInfo necessary?"
I honestly do not know, it's been years since I wrote this. Someone needs to do some research and testing.

And about your ideas, try what works best for yourself, no need to ask for permission on any change you want to make. I will intervene if a Pull Request has flaws and request changes.

creator3 commented:
Thanks, I didn't notice that it was cfg.save(*chunks*).

As far as I can tell, _integrated_removeBlockInfo isn't necessary because the same values will be written over with the new data anyway. I don't think removing it has broken anything on my server yet.

I've undone the version change, renamed the method to "store", and replaced "\u00a7r" with ChatColor.RESET.
